docs/source/federation/gateway.mdx

```diff
@@ -0,0 +1,227 @@
+});
+
+const server = new ApolloServer({
+  gateway,
+
+  // Disable subscriptions (not currently supported with ApolloGateway)
+  subscriptions: false,
+
+  context: ({ req }) => {
+    // get the user token from the headers
+    const token = req.headers.authorization || '';
+
+    // try to retrieve a user with the token
+    const userId = getUserId(token);
+
+    // add the user to the context
+    return { userId };
+  },
+});
+
+server.listen().then(({ url }) => {
+  console.log(`🚀 Server ready at ${url}`);
+});
+```
+
+</ExpansionPanel>
+
+The `buildService` function enables us to customize the requests that are sent to our implementing services. In this example, we return a custom `RemoteGraphQLDataSource`. The datasource allows us to modify the outgoing request with information from the Apollo Server `context` before it's sent. Here, we add the `user-id` header to pass an authenticated user ID to downstream services.
+
+### Customizing responses
+
+Let's say that whenever our implementing services return an operation result to the gateway, they include a `Server-Id` header in the response. The value of the header uniquely identifies the service in our graph.
+
+When the gateway then responds to a client, we want _its_ `Server-Id` header to include the identifier for _every_ service that contributed to the response. In this case, we can tell the gateway to aggregate the various server IDs into a single, comma-separated list.
+
+The flow for processing a single operation from a client application then looks like this:
+
+```mermaid
+sequenceDiagram;
+  Client app->>Gateway: Sends GraphQL operation
+  Gateway->>Gateway: Generates query plan for operation
+  loop For each operation in the query plan
+    Gateway->>Implementing Services: Sends the operation to the applicable service
+    Implementing Services->>Gateway: Responds with result and Server-Id header
+    Gateway->>Gateway: Adds the returned Server-Id to the shared context
+  end
+  Gateway->>Gateway: Adds all Server-Ids in the shared context to the response header
+  Gateway->>Client app: Sends operation response
+```
+
+To implement this flow, we can use the `didReceiveResponse` callback of the `RemoteGraphQLDataSource` class to inspect each implementing service's result as it comes in. We can add the `Server-Id` to the shared `context` in this callback, then pull the full list from the `context` when sending the final response to the client.
+
+<ExpansionPanel title="Expand example">
+
+```javascript
+const { ApolloServer } = require('apollo-server');
+const { ApolloGateway, RemoteGraphQLDataSource } = require('@apollo/gateway');
+
+class DataSourceWithServerId extends RemoteGraphQLDataSource {
+  // highlight-start
+  async didReceiveResponse({ response, request, context }) {
+    // Parse the Server-Id header and add it to the array on context
+    const serverId = response.http.headers.get('Server-Id');
+    if (serverId) {
+      context.serverIds.push(serverId);
```

Comment:
You've taken care in your implementation so this isn't a problem, though I feel a warning may be in order — similar to the warning which we (now?) give with regards to adjusting `context` within resolvers in the Resolvers page (after https://github.com/apollographql/apollo-server/pull/3834).  

In other words, guidance which advises to not make assumptions about the order of a services' `buildService` invocation, since that can change dynamically depending on the shape of the query and is determined by the query planner.

In other words, setting `context.prop` in one service is not necessarily guaranteed to happen before another service steps on it by doing the same thing.  Therefore, `push`-ing into, e.g., an Array, or an object keyed by the service's name is a more defensive approach (though while `buildService` receives the `name` of a service, the default `RemoteGraphQLDataSource` doesn't accept `name` itself. 🤔  Anyhow, I think my point stands and it could be keyed by something else as well).